pkg/handler/collector/oci/oci.go

```diff
@@ -0,0 +1,222 @@
+	rcOpts := []regclient.Opt{}
+	rcOpts = append(rcOpts, regclient.WithDockerCreds())
+	rcOpts = append(rcOpts, regclient.WithDockerCerts())
+
+	if o.tag != "" {
+		imageTag := fmt.Sprintf("%v:%v", o.repo, o.tag)
+		r, err := ref.New(imageTag)
+		if err != nil {
+			return err
+		}
+
+		rc := regclient.New(rcOpts...)
+		defer rc.Close(ctx, r)
+
+		err = o.fetchOCIArtifacts(ctx, rc, r, docChannel)
+		if err != nil {
+			return err
+		}
+	} else {
+		r, err := ref.New(o.repo)
+		if err != nil {
+			return err
+		}
+
+		rc := regclient.New(rcOpts...)
+		defer rc.Close(ctx, r)
+
+		tags, err := rc.TagList(ctx, r)
+		if err != nil {
+			return fmt.Errorf("reading tags for %s: %w", o.repo, err)
+		}
+
+		for _, tag := range tags.Tags {
+			if !strings.HasSuffix(tag, "sbom") && !strings.HasSuffix(tag, "att") && !strings.HasSuffix(tag, "sig") {
+				imageTag := fmt.Sprintf("%v:%v", o.repo, tag)
+				r, err := ref.New(imageTag)
+				if err != nil {
+					return err
+				}
+				err = o.fetchOCIArtifacts(ctx, rc, r, docChannel)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (o *ociCollector) fetchOCIArtifacts(ctx context.Context, rc *regclient.RegClient, image ref.Ref, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+
+	// attempt to request only the headers, avoids Docker Hub rate limits
+	m, err := rc.ManifestHead(ctx, image)
+	if err != nil {
+		return err
+	}
+
+	for m.IsList() {
+		pl, _ := manifest.GetPlatformList(m)
+		for _, p := range pl {
+			desc, err := manifest.GetPlatformDesc(m, p)
+			if err != nil {
+				return fmt.Errorf("failed retrieving platform specific digest: %w", err)
+			}
+			image.Digest = desc.Digest.String()
+			err = o.fetchOCIArtifacts(ctx, rc, image, docChannel)
+			if err != nil {
+				return fmt.Errorf("failed retrieving platform specific digest: %w", err)
+			}
+		}
+	}
+
+	digest := manifest.GetDigest(m)
+	digestFormatted := fmt.Sprintf("%v-%v", digest.Algorithm(), digest.Encoded())
+	// check to see if the digest has already been collected
+	if !contains(o.checkedDigest, digestFormatted) {
+		suffixList := []string{"att", "sbom"}
+		for _, suffix := range suffixList {
+			digestTag := fmt.Sprintf("%v.%v", digestFormatted, suffix)
+			imageTag := fmt.Sprintf("%v:%v", o.repo, digestTag)
+			r, err := ref.New(imageTag)
+			if err != nil {
+				return err
+			}
+
+			// if `.att` or `.sbom`` do not exist for specified digest
+			// log error and continue
+			m, err = rc.ManifestGet(ctx, r)
+			if err != nil {
+				logger.Error(err)
+				continue
+			}
+
+			// go through layers in reverse
+			mi, ok := m.(manifest.Imager)
+			if !ok {
+				return fmt.Errorf("reference is not a known image media type")
+			}
+			layers, err := mi.GetLayers()
+			if err != nil {
+				return err
+			}
+			for i := len(layers) - 1; i >= 0; i-- {
+				blob, err := rc.BlobGet(ctx, r, layers[i])
```

Comment:
You may want to enforce a size limit here, since a blob can grow rather large and you're reading into memory below.

Reply:
Currently, we don't enforce the size limit for the blobs we collect in any of the collectors. Maybe we might need to consider it.